### PR TITLE
Vrf reveal

### PR DESCRIPTION
**Pull request by Mathieu-Be**
Implementing VRF for Launchpeg. A few notes : 
- I kept `forceReveal()` with pseudo-randomness for safety
- Multisig will need to add the launchpeg contract to its list of VRF consumers.

An example on Fuji can be found [here](https://testnet.snowtrace.io/address/0x5932677b3f981d718114dc9b713beec51e2c9c08)

Not related but I noticed that the tests were forking mainnet. I removed it, I think it was unnecessarily slowing down the tests.

Detailed description of how this works [here](https://www.notion.so/traderjoexyz/Chainlink-VRF-for-Launchpeg-e60305b39c0b4ad086ec567a71fc602a).

### CODE REVIEW

**Review by cryptofish7 — CHANGES_REQUESTED** [review]

File: contracts/BatchReveal.sol [hunk 1]
```diff
@@ -255,13 +295,43 @@ abstract contract BatchReveal is IBatchReveal, Initializable {
...
-        _setBatchSeed(batchNumber);
+        if (useVRF) {
+            VRFCoordinatorV2Interface(vrfCoordinator).requestRandomWords(
+                keyHash,
+                subscriptionId,
+                requestConfirmations,
+                callbackGasLimit,
+                1
+            );
+            vrfRequestedForBatch[batchNumber] = true;
```
[the remark this answers is not in the document]
Comment (Mathieu-Be) [reply contracts/BatchReveal.sol:306]: The coordinator already emits an event on `requestRandomWords`, with all the VRF params and the address of the msg.sender

File: contracts/BatchReveal.sol [hunk 2]
```diff
@@ -255,13 +295,43 @@ abstract contract BatchReveal is IBatchReveal, Initializable {
...
+        if (hasBeenForceRevealed) {
+            revert Launchpeg__HasBeenForceRevealed();
+        }
+        lastTokenRevealed += revealBatchSize;
+        batchToSeed[nextBatchToReveal] =
+            randomWords[0] %
+            (collectionSize - (nextBatchToReveal * revealBatchSize));
+
+        emit Reveal(nextBatchToReveal, batchToSeed[nextBatchToReveal]);
+        nextBatchToReveal++;
```
[the remark this answers is not in the document]
Comment (Mathieu-Be) [reply contracts/BatchReveal.sol]: Does the `++` on `nextBatchToReveal` happens before or after being initializing `_batchToReveal` ?

File: contracts/BaseLaunchpeg.sol [hunk 3]
```diff
@@ -286,6 +298,31 @@ abstract contract BaseLaunchpeg is
...
+    /// @param _vrfCoordinator Chainlink coordinator address
+    /// @param _keyHash Keyhash of the gas lane wanted
+    /// @param _subscriptionId Chainlink subscription ID
+    /// @param _callbackGasLimit Max gas used by the coordinator callback
+    function initializeVRF(
+        address _vrfCoordinator,
+        bytes32 _keyHash,
+        uint64 _subscriptionId,
+        uint32 _callbackGasLimit
+    ) external onlyOwner {
```
[the remark this answers is not in the document]
Comment (Mathieu-Be) [reply contracts/BaseLaunchpeg.sol:311]: I went for the first option as I see no reasons to lock the parameters. Also added some validation

File: contracts/BatchReveal.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (LouisMeMyself) [reply contracts/BatchReveal.sol]: After, that's why it's so elegant!

If you wanted to do it before, it would be `++nextBatchToReveal`.

**Comment by Mathieu-Be** [comment]
Deployed a new contract on Fuji : https://testnet.snowtrace.io/address/0xa8e171869fc5a8da7183d535906d0fca76a51b41

File: contracts/BatchReveal.sol [hunk 4]
```diff
@@ -37,6 +39,41 @@ abstract contract BatchReveal is IBatchReveal, Initializable {
...
+
+    /// @notice Depends on the number of requested values that you want sent to the
+    /// fulfillRandomWords() function. Storing each word costs about 20,000 gas,
+    /// so 100,000 is a safe default for this example contract. Test and adjust
+    /// this limit based on the network that you select, the size of the request,
+    /// and the processing of the callback request in the fulfillRandomWords()
+    /// function.
+    uint32 public override callbackGasLimit;
+
+    /// @notice Number of block confirmations that will the coordinator wait before triggering the callback
```
[the remark this answers is not in the document]
Comment (cryptofish7) [reply contracts/BatchReveal.sol:61]: `Number of block confirmations that the coordinator will wait before triggering the callback`

**Review by cryptofish7 — CHANGES_REQUESTED** [review]
Small nit

File: contracts/BaseLaunchpeg.sol [hunk 5]
```diff
@@ -314,14 +314,43 @@ abstract contract BaseLaunchpeg is
...
 
         // 20_000 is the cost of storing one word, callback cost will never be lower than that
         if (_callbackGasLimit > _maxGasLimit || _callbackGasLimit < 20_000) {
             revert Launchpeg__InvalidCallbackGasLimit();
         }
 
+        bool keyHashFound;
+        for (uint256 i; i < s_provingKeyHashes.length; i++) {
+            if (s_provingKeyHashes[i] == _keyHash) {
+                keyHashFound = true;
```
Comment (LouisMeMyself) [inline contracts/BaseLaunchpeg.sol:331]: Add `break` when found

File: contracts/BaseLaunchpeg.sol [hunk 6]
```diff
@@ -314,14 +314,43 @@ abstract contract BaseLaunchpeg is
...
+        }
+
+        (, , , address[] memory consumers) = VRFCoordinatorV2Interface(
+            _vrfCoordinator
+        ).getSubscription(_subscriptionId);
+
+        bool isInConsumerList;
+        for (uint256 i; i < consumers.length; i++) {
+            if (consumers[i] == address(this)) {
+                isInConsumerList = true;
```
Comment (LouisMeMyself) [inline contracts/BaseLaunchpeg.sol:346]: Add `break` when found

**Review by LouisMeMyself — CHANGES_REQUESTED** [review]
Good stuff!

**Review by cryptofish7 — APPROVED** [review]

**Pull request merged**